src/sidebar/store/modules/annotations.js

```diff
@@ -330,19 +321,42 @@ function hideAnnotation(id) {
 }
 
 /**
- * Create a new annotation
+ * Create a new annotation (as-yet unpersisted)
+ *
+ * This method has several responsibilities:
+ * 1. Set some default data attributes on the annotation
+ * 2. Remove any existing, empty drafts
+ * 3. Add the annotation to the current collection of annotations
+ * 4. Change focused tab to the applicable one for the new annotation's meta-type
+ * 5. Expand all of the new annotation's parents
  *
- * The method does 4 tasks:
- * 1. Removes any existing empty drafts.
- * 2. Creates a new annotation.
- * 3. Changes the focused tab to match that of the newly created annotation.
- * 4. Expands the collapsed state of all new annotation's parents.
  */
-function createAnnotation(ann) {
-  return dispatch => {
+function createAnnotation(ann, now = new Date()) {
+  return (dispatch, getState) => {
+    /**
+     * Extend the new, unsaved annotation object with defaults for some
+     * required data fields.
+     *
+     * Note: the `created` and `updated` values will be ignored and superseded
+     * by the service when the annotation is persisted, but they are used
+     * app-side for annotation card sorting until then.
+     */
+    ann = Object.assign(
```

Comment:
My coding gut really wants to abstract this logic into a function, but the two options I had here didn't seem ideal:

1. Passing `getState` to said function so that it can derive the values it needs—passing `getState` around may likely violate conventions/be a bad idea, or,
1. Deriving the state-dependent values needed by the function and passing them as arguments. Ended up with an argument list some six or seven long, and twice the LOC.

Very open to suggestions here.

Reply:
> My coding gut really wants to abstract this logic into a function

Unless you have a clear idea of what it is that you want to abstract, I think this would just add indirection.

The main potential issue I see is something different - someone could change the state shape of the groups or session modules and not realize that this module depends upon it.